src/GRANULAR/granular_model.cpp

```diff
@@ -261,9 +261,9 @@ void GranularModel::init()
     if (sub_models[i]->beyond_contact)
       beyond_contact = 1;
     size_history += sub_models[i]->size_history;
-    if (!sub_models[i]->allow_cohesion && normal_model->get_cohesive_flag())
-      error->all(FLERR,"Cannot use {} model with a cohesive normal model, {}",
-                 sub_models[i]->name, normal_model->name);
+    // if (!sub_models[i]->allow_cohesion && normal_model->get_cohesive_flag())
+    //   error->all(FLERR,"Cannot use {} model with a cohesive normal model, {}",
+    //              sub_models[i]->name, normal_model->name);
```

Comment:
Was this an intended change?

Reply:
Oh no. Sorry, some of my labmates wanted to use  coeff_resitution model with cohesive models. So I changed this for them.

Reply:
In this updated version, I’ve introduced a toggle called "synchronized_verlet", which works similarly to the existing "limit_damping" toggle. Additionally, I added a rotate_rescale_vec function to manage the rotation and rescaling of history and velocity terms.

As suggested by you I’m considering creating a distinct subclass that activates when "synchronized_verlet" is enabled. This subclass would be responsible for handling tangential linear history and Mindlin forces, while the standard subclass would apply when "synchronized_verlet" is not active. However, I was not sure how to use the toggle to change between these. Could you provide guidance on the best approach to implement this?

Reply:
> I’m curious, though—why doesn’t it work for cohesive models? In the absence of cohesion (when γ = 0), the model should theoretically exhibit Hertzian-like behavior, so wouldn’t it be reasonable to use this damping model with the condition $ \delta > 0$?

Since JKR is intended to be used with gamma != 0, where a connection between an input and output CoR is unknown, it was blocked to avoid misleading users who don't look into the numerical details of their contact model choices. Similarly Tsuji errors since that derivation is based on no cohesion (if I recall correctly). Originally, both would also produce nan's with negative overlaps (now fixed to guard against rare machine precision issues, but discussed here https://github.com/lammps/lammps/issues/4180). 

Of course, one can always turn off the error like your lab mate did. The error could also be turned into a warning that prints more details, but then the documentation would have to be updated to clearly describe how these damping forces work in this scenario and contextualize the choice. So far, no one has taken the time to argue for this change and make the necessary changes.

Reply:
> In this updated version, I’ve introduced a toggle called "synchronized_verlet", which works similarly to the existing "limit_damping" toggle. Additionally, I added a rotate_rescale_vec function to manage the rotation and rescaling of history and velocity terms.

Great, that sounds like it should work well and (hopefully) shouldn't be much effort to document. I'll try to take a look at the code tomorrow. 

> As suggested by you I’m considering creating a distinct subclass that activates when "synchronized_verlet" is enabled. This subclass would be responsible for handling tangential linear history and Mindlin forces, while the standard subclass would apply when "synchronized_verlet" is not active. However, I was not sure how to use the toggle to change between these subclasses. Could you provide guidance on the best approach to implement this?

I was originally thinking this solution would be an alternate implementation without adding a "synchronized_verlet" model. But maybe they could somehow work together. Let me think about it a bit.

Reply:
With more thought, I think automatically picking different submodel variants is a bad idea. Implementing it would be a bit clunky and it feels more opaque. I think it'd be better to have variants that work with the alternate integration scheme and have LAMMPS error if you mismatch a submodel variant with your synchonized verlet flag.

Reply:
> > I’m curious, though—why doesn’t it work for cohesive models? In the absence of cohesion (when γ = 0), the model should theoretically exhibit Hertzian-like behavior, so wouldn’t it be reasonable to use this damping model with the condition $ \delta > 0$?
> 
> Since JKR is intended to be used with gamma != 0, where a connection between an input and output CoR is unknown, it was blocked to avoid misleading users who don't look into the numerical details of their contact model choices. Similarly Tsuji errors since that derivation is based on no cohesion (if I recall correctly). Originally, both would also produce nan's with negative overlaps (now fixed to guard against rare machine precision issues, but discussed here #4180).
> 
> Of course, one can always turn off the error like your lab mate did. The error could also be turned into a warning that prints more details, but then the documentation would have to be updated to clearly describe how these damping forces work in this scenario and contextualize the choice. So far, no one has taken the time to argue for this change and make the necessary changes.

Makes sense. Thank you.

Reply:
> With more thought, I think automatically picking different submodel variants is a bad idea. Implementing it would be a bit clunky and it feels more opaque. I think it'd be better to have variants that work with the alternate integration scheme and have LAMMPS error if you mismatch a submodel variant with your synchonized verlet flag.

Could you please expand on what you mean by having a variant and how one would go about implementing it?

Reply:
I mean  a version of some submodels like *mindlin_synchronized* that include your changes, leaving the original *mindlin* umodified. 

I presume this is what you were referring to when you mentioned "distinct subclass that activates when 'synchronized_verlet' is enabled." So activating, e.g., *mindlin_synchronized* instead of *mindlin*. But maybe I misunderstood.

Reply:
Got it. So, when the `synchronised_verlet` flag is enabled, the code will call either `mindlin_synchronized` or `tangential_linear_synchronized` based on the selected model. If a new model without a synchronized version tries to use the `synchronised_verlet` flag, it will trigger an error for the user. Is that the idea?

Reply:
Well, I don't think the code should automatically swap the model. I think the two options are (correct me if I'm misunderstanding):

1) The flag changes the behavior of existing submodels (like you currently have it, no new submodels only new conditional logic)
2) Create new submodels, the flag blocks use of incompatible submodels.

Since you already have (1) implemented, I would just stick with that unless there's another reason for (2)?

Reply:
Alright, in that case, it looks like what we have already works! 

Do you want me to make any further changes to the code, or should I start working on test cases and documentation?

Reply:
I think the code only needs minor edits (cleaning up comments and fixing the style issues). It might be good to still add a *allows_synchronization* flag to the submodel classes to confirm they are compatible (maybe future user-contributed tangential/rolling models might not have it). But that can also be done down the road if a conflict emerges.

Documentation + examples are definitely the most important additions at this point.

Reply:
The `allows_synchronization` flag is a great idea! I’ll get started on this and will keep you updated when it's ready. I'm traveling next week, so it might take a bit of time, but I’ll aim to finish before then.